### ISSUE TITLE
Cordova is rejecting Java and gradle on windows

### ISSUE THREAD

**Issue by ghost** [apache/cordova-android#1554]
hi, total noob. this is my first question on a github ever in >10 years. 
im getting this error that chat gpt seems to think is associated with either android version 61 or jdk-17.0.2
its not clear if 
`cordova platform list`
`Installed platforms:
  android 10.1.2`
  supports versions over JDK 10 or really what those are despite 3 workdays worth of chatGPT and googling.
i have downloaded JDK 8 and 19 and 11 but im not quite sure if 8 comes before 1.8 or after it because some times its called JDK18 and sometimes 1.8 so its quite confusing. currently i am using JDK17 as mentioned, but chatGPT seems to think i should dual install it with version 11. I think it's hallucinating, but I can't make sense of what this post is trying to say RE java versions either https://stackoverflow.com/questions/40628240/cordova-requirements-check-failed-for-jdk-1-8-or-greater

that said, I am 75% sure its not a java issue, but a gradle issue due to the error message i am getting.

Can anyone provide some context to this?
OS: windows 10
fresh android studio install
working on non OS drive 
`>cordova -v
11.1.0`

```cordova run
Checking Java JDK and Android SDK versions
ANDROID_SDK_ROOT=\SDK (recommended setting)
ANDROID_HOME=undefined (DEPRECATED)
Using Android SDK: \SDK
Subproject Path: CordovaLib
Subproject Path: app

FAILURE: Build failed with an exception.

* Where:
Settings file '\Cordova\platforms\android\settings.gradle'

* What went wrong:
Could not compile settings file '\Cordova\platforms\android\settings.gradle'.
> startup failed:
  General error during conversion: Unsupported class file major version 61```
followed by 
```
 1 error

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.1.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 926ms```

**Comment by breautek** [comment]
First some background knowledge:

cordova-android 10 if you were following the documentation, uses the legacy Android SDK parts which expects Java 8, but *is* compatible with Java 11, if your environment is configured to use the modern SDK tooling (e.g. using `cmdline-tools` instead of `tools/` found in the platform package).

cordova-android@11 to support API 31 and later requires the use of current Android SDK tooling which expects Java 11.

cordova-android@9 and earlier used to check and enforce that you're running Java 8. That check was removed in version 10, so your SO link is in referencing to a vesion older than 10, and is out of date now.

Note that the `cordova` version doesn't necessarily match with the platform (cordova-android) version.
Use `cordova platform ls` to list platform versions.

> but im not quite sure if 8 comes before 1.8 

1.8 is Java 8. Java historically used 1.0, 1.1, 1.2, .. so on. They switched to a simple number at version 5, so it was Java 5, instead of Java 1.5. However, some people continued to reference the versions using the old name scheme.

> im getting this error that chat gpt seems to think is associated with either android version 61 or jdk-17.0.2

"file major version 61" indicates you're using Java 17.x, which isn't a recommended version.

> that said, I am 75% sure its not a java issue, but a gradle issue due to the error message i am getting.

The error does indicate a java issue. Gradle is essentially saying it isn't compatible with file major version 61, or JDK 17.

My recommendation:

Keep up with versions. cordova 11 installs cordova-android 10 by default, but there is a newer version available, so if possible upgrade to cordova-android@11. This can be done by doing:

```
cordova platform remove android
cordova platform add android@11
```

If you plan on deploying to google play, cordova-android@11 is required.

For cordova projects, use Java/JDK 11. This is what Android Studio uses and using newer JDK versions may not work. Android Studio does ship JDK with itself, so you can point it to that installation (should be inside `<android-studio-root>/jre`, or you can install JDK 11 from your favourite vendor.

**Comment by 0T0K0-Tensh1** [comment]
that info was somewhat helpful but didn't resolve my issue.
tl;dr post below, please read it first.
1st-ly,
the `<android-studio-root>/jre` path does not contain anything but an empty file named .marker

2nd

`_\Cordova>_cordova requirements android`
now gives 
`Android target: not installed
Command failed with exit code 1: avdmanager list target`
like so:
```
_\Cordova>_cordova requirements android

Requirements check results for android:
Java JDK: installed 11.0.18
Android SDK: installed true
**Android target: not installed**
**Command failed with exit code 1: avdmanager list target**
Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema
        at com.android.repository.api.SchemaModule$SchemaModuleVersion.<init>(SchemaModule.java:156)
        at com.android.repository.api.SchemaModule.<init>(SchemaModule.java:75)
        at com.android.sdklib.repository.AndroidSdkHandler.<clinit>(AndroidSdkHandler.java:81)
        at com.android.sdklib.tool.AvdManagerCli.run(AvdManagerCli.java:213)
        at com.android.sdklib.tool.AvdManagerCli.main(AvdManagerCli.java:200)
Caused by: java.lang.ClassNotFoundException: javax.xml.bind.annotation.XmlSchema
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 5 more
Gradle: installed \AndroidDev\gradle-7.6\bin\gradle.BAT
Some of requirements check failed
```

i also tried to import to my quasar project but got a different error talking about gradle and version 8? it seems to be able to find the apk that cordova cant for some reason. but I'm not sure if it is cordova or quasar building it. afaik it has to be cordova since quasar is using it in the first place?

(the venv is an unrelated python environment for script commands.)

if true, it made me think that there is a problem with cordova's handling of params.
running the commands above and below only work in the cordova environment i defined in a .bat file (see below). without that, quasar works fine, but cannot boot cordova. 


```
quasar mode add cordova
 App • ⚠  Cordova support detected already. Aborting.

_\quasar-project>_cordova platform ls
Current working directory is not a Cordova-based project.

_\quasar-project>_quasar dev --mode cordova --target android

 .d88888b.
d88P" "Y88b
888     888
888     888 888  888  8888b.  .d8888b   8888b.  888d888
888     888 888  888     "88b 88K          "88b 888P"
888 Y8b 888 888  888 .d888888 "Y8888b. .d888888 888
Y88b.Y8b88P Y88b 888 888  888      X88 888  888 888
 "Y888888"   "Y88888 "Y888888  88888P' "Y888888 888
       Y8b

? What external IP should Quasar use? (Use arrow keys)
App • Updated Cordova config.xml
 App • Running "cordova run android" in \AndroidDev\Apps\Quasar\quasar-project\src-cordova

Checking Java JDK and Android SDK versions
ANDROID_SDK_ROOT=<path-to-sdk> (recommended setting)
ANDROID_HOME=\AndroidDev\SDK (DEPRECATED)
Using Android SDK: \AndroidDev\SDK

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
> Starting Daemon
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.6/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 25s
1 actionable task: 1 executed
Subproject Path: CordovaLib
Subproject Path: app
Starting a Gradle Daemon (subsequent builds will be faster)
Warning: Mapping new ns http://schemas.android.com/repository/android/common/02 to old ns http://schemas.android.com/repository/android/common/01
...
Warning: unexpected element (uri:"", local:"base-extension"). Expected elements are <{}codename>,<{}layoutlib>,<{}api-level>

> Task :CordovaLib:compileDebugJavaWithJavac
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.1.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 1m 31s
48 actionable tasks: 48 executed
**Built the following apk(s):**
        \AndroidDev\Apps\Quasar\quasar-project\src-cordova\platforms\android\app\build\outputs\apk\debug\app-debug.apk
Checking Java JDK and Android SDK versions
ANDROID_SDK_ROOT=\AndroidDev\SDK (recommended setting)
ANDROID_HOME=\AndroidDev\SDK (DEPRECATED)
Using Android SDK: \AndroidDev\SDK
Command failed with exit code 1: apkanalyzer manifest target-sdk \AndroidDev\Apps\Quasar\quasar-project\src-cordova\platforms\android\app\build\outputs\apk\debug\app-debug.apk
**'sh' is not recognized as an internal or external command,**
operable program or batch file.

 App • Command "cordova" failed with exit code: 1
 App • Updated Cordova config.xml

 App • ⚠   FAIL  Cordova CLI has failed
```
 
 it also asked me "what is the cordova id" and i just hit enter, but the next 2 times (2nd time in cordova environment above) i didn't get that option, so im wondering if i need to define it.


the bat looks like this
```@echo off
set "ANDROID_HOME=D:\AndroidDev\SDK"
set "JAVA_HOME=C:\Program Files\Java\jdk-11"
set "GRADLE_HOME=D:\AndroidProject\.gradle"

start "" /D"D:\Cordova\" /B cmd.exe /k D:\Cordova\venv_\Scripts\activate.bat /B 
cls
```
.bat gives these vars when asked
```
ANDROID_SDK_ROOT:
D:\AndroidDev\SDK
JAVA_HOME:
C:\Program Files\Java\jdk-11
GRADLE_HOME: D:\AndroidProject\.gradle
```
even with the GRADELE_HOME unset, like so: 
![image](https://user-images.githubusercontent.com/6926311/216847523-4fb6b8a4-b5ae-4847-9e90-f8bb9a6ebc7f.png)
it changed nothing

**Comment by 0T0K0-Tensh1** [comment]
### _tl;dr_
@breautek 

i am still getting errors such as this

**quasar dev --mode cordova --target android** gives
`**'sh' is not recognized as an internal or external command,**`
`Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.`

_>_ cordova requirements android
```
Requirements check results for android:
Java JDK: installed 11.0.18
Android SDK: installed true
Android target: not installed
```

**Comment by breautek** [comment]
I can't really answer for third-party CLI tools, I don't use them.

```
**Command failed with exit code 1: avdmanager list target**
Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema
        at com.android.repository.api.SchemaModule$SchemaModuleVersion.<init>(SchemaModule.java:156)
        at com.android.repository.api.SchemaModule.<init>(SchemaModule.java:75)
        at com.android.sdklib.repository.AndroidSdkHandler.<clinit>(AndroidSdkHandler.java:81)
        at com.android.sdklib.tool.AvdManagerCli.run(AvdManagerCli.java:213)
        at com.android.sdklib.tool.AvdManagerCli.main(AvdManagerCli.java:200)
Caused by: java.lang.ClassNotFoundException: javax.xml.bind.annotation.XmlSchema
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 5 more
```

I'm fairly certain this is the error you get if you mismatch Android SDK tooling with java versions. The Android SDK includes two sets of the same tools, once is for Java 8, the other is for Java 11.

The java 8 tooling I'm pretty sure is automatically bundled with the android SDK under `$ANDROID_HOME/tools` / `$ANDROID_HOME/tools/bin`. So ensure these paths are **not** in your `PATH` variable, as they will not work if you're using Java 11.

Instead, using the SDK Manager, install the `cmdline-tools` package. I'm using version 7.0 right now just fine. Then you can add `$ANDROID_HOME/cmdline-tools/7.0/bin` to your `PATH`. These binaries are compatible with Java 11. If you use a different version, make sure you adjust your PATH as the version is in the path.

> **'sh' is not recognized as an internal or external command,**

I'm guessing it's trying to start the emulator and is attempting to run a sh command on the emulator, but if the `avdmanager` binaries aren't working because of the mismatch of java versions, it will fail. So this is probably a side effect error of your first problem.

Lastly, I know you're using Windows, so my `$ANDROID_HOME/tools` examples are not going to be identical as they are based on unix environments... I don't develop on a windows machine, so I'm not very familiar with how they represent things. I'm assuming that you have the know how to set up environment variables on Windows. I think windows uses `%ANDROID_HOME%` for environment variables. `$ANDROID_HOME` should be the path to the root android sdk folder.

> the <android-studio-root>/jre path does not contain anything but an empty file named

Perhaps windows has a different folder structure. To be clear this is not the Android SDK folder, but the Android Studio folder. But on Linux you download Android Studio as one big tarball archive whereas on windows you get an actual installer, so the structure could be very different.

**Comment by 0T0K0-Tensh1** [comment]
didnt help
i just used it with quantum and it worked
no idea why
anyway im deleting my github now

**Comment by kvker** [comment]
it is helpful.
"Instead, using the SDK Manager, install the cmdline-tools package. I'm using version 7.0 right now just fine. Then you can add $ANDROID_HOME/cmdline-tools/7.0/bin to your PATH. These binaries are compatible with Java 11. If you use a different version, make sure you adjust your PATH as the version is in the path."

tks.

**Comment by gotohigher** [comment]
cordova requirements

Requirements check results for android:
Java JDK: installed 11.0.2
Android SDK: installed true
Android target: not installed
Command failed with exit code 1: avdmanager list target
Error occurred during initialization of boot layer
java.lang.module.FindException: Module java.xml.bind not found

how do I fix it?

**Comment by breautek** [comment]
> cordova requirements
> 
> Requirements check results for android: Java JDK: installed 11.0.2 Android SDK: installed true Android target: not installed Command failed with exit code 1: avdmanager list target Error occurred during initialization of boot layer java.lang.module.FindException: Module java.xml.bind not found
> 
> I've already set JAVA_HOME, ANDROID_ROOT and jdk path how do I fix it?

> Command failed with exit code 1: avdmanager list target

This is a bug and will always fail currently. It shouldn't prevent you from actually running cordova.

> Error occurred during initialization of boot layer
java.lang.module.FindException: Module java.xml.bind not found

Java exception issues when using android sdk tools are usually always a result of mixing newer tools with older tools.

At this point I'd recommend using cordova-android@12 (or 13 when it's released) with JDK 17, the latest command line tools available. Make sure platform tools, build tools, and emulator is also all updated to the latest versions. If you have `Android SDK Tools (Obsolete)` installed, then remove uninstall it.

See https://github.com/apache/cordova/discussions/475 for more details

**Comment by gotohigher** [comment]
cordova -version
12.0.0 (cordova-lib@12.0.1)

cordova platform add android@12.0.1
Using cordova-fetch for cordova-android@12.0.1
Adding android project...
Creating Cordova project for the Android platform:
        Path: platforms\android
        Package: com.example.hello
        Name: HelloWorld
        Activity: MainActivity
        Android Target SDK: android-33
        Android Compile SDK: 33
Subproject Path: CordovaLib
Subproject Path: app
Android project created with cordova-android@12.0.1


I updated JDK 17, and latest platform tools, build tools and emulator. but result is same

**Comment by lewismunene020** [comment]
i am also facing  the same  issue on my end  what might  be  the problem

**Comment by RajnishAnand** [comment]
depending on the version installed its required to add `$ANDROID_HOME/cmdline-tools/<version>/bin` to path
for me it was `$ANDROID_HOME/cmdline-tools/latest/bin` . 
and just to be clear `ANDROID_HOME=C:\Users\<username>\AppData\Local\Android\Sdk\` which requires to be set as environmental variable.

**Comment by lewismunene020** [comment]
i also  did an upgrade  of  the gradle  version and it  worked pretty well

**Comment by peteshand** [comment]
Is Cordova still actively maintained?

I've been attempting to get a basic sample app running, but every combination of Cordova, Android SDK, Gradle, and Java results in different errors. Is there a definitive list of compatible versions for these tools?

**Comment by peteshand** [comment]
Gradle <-> Java version compatibility:

https://docs.gradle.org/current/userguide/compatibility.html?_ga=2.123391874.842902488.1720258672-1831996987.1720258672&_gl=1*12b3vyt*_gcl_au*NTQwMTA4MDM4LjE3MjAyNTg2NzI.*_ga*MTgzMTk5Njk4Ny4xNzIwMjU4Njcy*_ga_7W7NC6YNPT*MTcyMDI1ODY3Mi4xLjAuMTcyMDI1ODY3Mi42MC4wLjA.

**Comment by fattoliu** [comment]
> I can't really answer for third-party CLI tools, I don't use them.我无法真正回答第三方 CLI 工具的问题，我不使用它们。
> 
> ```
> **Command failed with exit code 1: avdmanager list target**
> Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema
>         at com.android.repository.api.SchemaModule$SchemaModuleVersion.<init>(SchemaModule.java:156)
>         at com.android.repository.api.SchemaModule.<init>(SchemaModule.java:75)
>         at com.android.sdklib.repository.AndroidSdkHandler.<clinit>(AndroidSdkHandler.java:81)
>         at com.android.sdklib.tool.AvdManagerCli.run(AvdManagerCli.java:213)
>         at com.android.sdklib.tool.AvdManagerCli.main(AvdManagerCli.java:200)
> Caused by: java.lang.ClassNotFoundException: javax.xml.bind.annotation.XmlSchema
>         at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
>         at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
>         at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
>         ... 5 more
> ```
> 
> I'm fairly certain this is the error you get if you mismatch Android SDK tooling with java versions. The Android SDK includes two sets of the same tools, once is for Java 8, the other is for Java 11.我相当确定，如果你将 Android SDK 工具与 Java 版本不匹配，你会得到这个错误。Android SDK 包含两套相同的工具，一套是为 Java 8 准备的，另一套是为 Java 11 准备的。
> 
> The java 8 tooling I'm pretty sure is automatically bundled with the android SDK under `$ANDROID_HOME/tools` / `$ANDROID_HOME/tools/bin`. So ensure these paths are **not** in your `PATH` variable, as they will not work if you're using Java 11.我相当确信，Java 8 工具是在 android SDK 中自动捆绑的，位于 `$ANDROID_HOME/tools` / `$ANDROID_HOME/tools/bin` 。所以请确保这些路径不在你的 `PATH` 变量中，因为如果你使用 Java 11，它们将无法工作。
> 
> Instead, using the SDK Manager, install the `cmdline-tools` package. I'm using version 7.0 right now just fine. Then you can add `$ANDROID_HOME/cmdline-tools/7.0/bin` to your `PATH`. These binaries are compatible with Java 11. If you use a different version, make sure you adjust your PATH as the version is in the path.相反，使用 SDK Manager 安装 `cmdline-tools` 包。我现在使用的是版本 7.0，效果很好。然后你可以将 `$ANDROID_HOME/cmdline-tools/7.0/bin` 添加到你的 `PATH` 。这些二进制文件与 Java 11 兼容。如果你使用不同的版本，请确保调整你的 PATH，因为版本就在路径中。
> 
> > **'sh' is not recognized as an internal or external command,'sh'不是内部或外部命令，**
> 
> I'm guessing it's trying to start the emulator and is attempting to run a sh command on the emulator, but if the `avdmanager` binaries aren't working because of the mismatch of java versions, it will fail. So this is probably a side effect error of your first problem.我猜它正在尝试启动模拟器，并尝试在模拟器上运行 sh 命令，但如果由于 Java 版本不匹配， `avdmanager` 的二进制文件无法工作，它将失败。所以这很可能是你第一个问题的副作用错误。
> 
> Lastly, I know you're using Windows, so my `$ANDROID_HOME/tools` examples are not going to be identical as they are based on unix environments... I don't develop on a windows machine, so I'm not very familiar with how they represent things. I'm assuming that you have the know how to set up environment variables on Windows. I think windows uses `%ANDROID_HOME%` for environment variables. `$ANDROID_HOME` should be the path to the root android sdk folder.最后，我知道你在使用 Windows，所以我的 `$ANDROID_HOME/tools` 示例不会完全相同，因为它们是基于 Unix 环境的...我不在 Windows 机器上开发，所以我对它们如何表示事物不太熟悉。我假设你知道如何在 Windows 上设置环境变量。我认为 Windows 使用 `%ANDROID_HOME%` 作为环境变量。 `$ANDROID_HOME` 应该是根 Android SDK 文件夹的路径。
> 
> > the /jre path does not contain anything but an empty file named/jre 路径不包含任何内容，只有一个名为
> 
> Perhaps windows has a different folder structure. To be clear this is not the Android SDK folder, but the Android Studio folder. But on Linux you download Android Studio as one big tarball archive whereas on windows you get an actual installer, so the structure could be very different.可能 Windows 的文件夹结构不同。为了明确，这并不是 Android SDK 文件夹，而是 Android Studio 文件夹。但在 Linux 上，您下载 Android Studio 时是一个大的 tarball 存档，而在 Windows 上，您得到的是一个实际的安装程序，所以结构可能非常不同。

Thanks!
